### PR TITLE
Refactoring.pwm motor

### PR DESCRIPTION
**Pull request by bjornborg**
Pwm motor module

### CODE REVIEW

**Comment by bjornborg** [comment]
Please review my code and approve if it looks fine.

File: code/proxy-miniature/pwm-motor/src/PwmMotor.cpp [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+    return value;
+  } else {
+    cerr << "[" << getName() << "] Could not open " << filename 
+        << "." << std::endl;
+    file.close();
+    return 0;
+  }
+}
+
+uint32_t PwmMotor::Power2Pwm(double const a_percent) const
```
Comment (Remsworn) [inline code/proxy-miniature/pwm-motor/src/PwmMotor.cpp]: do not use "percent" in variable name if it is a fraction

File: code/proxy-miniature/pwm-motor/src/PwmMotor.cpp [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+  }
+  uint32_t pwm = static_cast<uint32_t>(std::round(m_minimumDutyCycleNs 
+      + a_percent * (m_maximumDutyCycleNs - m_minimumDutyCycleNs)));
+  return pwm;
+}
+
+double PwmMotor::Pwm2Power(uint32_t const a_pwm) const {
+  double power = (static_cast<double>(a_pwm) - m_minimumDutyCycleNs) 
+      / (m_maximumDutyCycleNs - m_minimumDutyCycleNs);
+  if (power < 0) {
```
Comment (Remsworn) [inline code/proxy-miniature/pwm-motor/src/PwmMotor.cpp]: what is the idea behind -1, serves a purpose?

File: usecases/latest/test.pwmmotor/configuration [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+###############################################################################
+#
+# GLOBAL CONFIGURATION
+#
+
+# The following attributes define the buffer sizes for recording and
+# replaying. You need to adjust these parameters depending on the
+# camera resolution for example (640x480x3 --> 1000000 for memorySegment,
+# 1280x720x3 --> 2800000).
+global.buffer.memorySegmentSize = 2800000 # Size of a memory segment in bytes.
```
Comment (Remsworn) [inline usecases/latest/test.pwmmotor/configuration]: camera res i pwm?

File: usecases/latest/test.pwmmotor/docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        - .:/opt/openkorp.data
+        command: "/opt/od4/bin/odsupercomponent --cid=${CID} --verbose=1 --configuration=/opt/openkorp.data/configuration"
+
+    proxy-miniature-pwm-motor-1:
+        build: .
+        network_mode: "host"
+        privileged: true
+        volumes:
+        - /sys/:/sys/
+        command: "/opt/openkorp/bin/opendlv-proxy-miniature-pwm-motor --cid=${CID} --freq=1 --id=1"
```
Comment (Remsworn) [inline usecases/latest/test.pwmmotor/docker-compose.yml]: frekvensen är rätt låg 1Hz ?

File: resources/odvd/ODVDMiniature.odvd [hunk 5]
```diff
@@ -39,6 +39,11 @@ message opendlv.proxy.PwmRequest [id = 155] {
   uint32 dutyCycleNs [id = 2];
 }
 
+message opendlv.proxy.PwmMotorRequest [id = 1000] {
```
Comment (Remsworn) [inline resources/odvd/ODVDMiniature.odvd]: q:What is this part for?

File: code/proxy-miniature/pwm-motor/src/PwmMotor.cpp [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+
+PwmMotor::~PwmMotor() 
+{
+}
+
+void PwmMotor::setUp()
+{
+  odcore::base::KeyValueConfiguration kv = getKeyValueConfiguration();
+
+  m_debug = (kv.getValue<int32_t>("proxy-miniature-pwm-motor.debug") == 1);
```
Comment (Remsworn) [inline code/proxy-miniature/pwm-motor/src/PwmMotor.cpp]: q: debugmode set from keyValueConfiguration file?

File: code/proxy-miniature/pwm-motor/man/opendlv-proxy-miniature-pwm-motor.1 [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+
+.SH SYNOPSIS
+.B opendlv-proxy-miniature-pwm-motor --cid=<CID>
+
+
+.SH EXAMPLES
+The following command joins the container conference 111:
+
+.B opendlv-proxy-miniature-pwm-motor --cid=111
```
Comment (Remsworn) [inline code/proxy-miniature/pwm-motor/man/opendlv-proxy-miniature-pwm-motor.1]: q:what is this file?

File: code/proxy-miniature/pwm-motor/src/PwmMotor.cpp [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+
+  m_initialised = true;
+}
+
+void PwmMotor::tearDown()
+{
+  SetDutyCycleNs(0);
+  SetEnabled(false);
+  SetPeriodNs(0);
+  ClosePwmMotor();
```
Comment (Remsworn) [inline code/proxy-miniature/pwm-motor/src/PwmMotor.cpp]: m_initialised = false; 
q: by the time of tearDown local variables to the components.

File: code/proxy-miniature/pwm-motor/src/PwmMotor.cpp [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+      SetDutyCycleNs(Power2Pwm(power));
+    }
+  }
+}
+
+odcore::data::dmcp::ModuleExitCodeMessage::ModuleExitCode PwmMotor::body()
+{
+  while (getModuleStateAndWaitForRemainingTimeInTimeslice() == 
+      odcore::data::dmcp::ModuleStateMessage::RUNNING) {
+    bool enabled = GetEnabled();
```
Comment (Remsworn) [inline code/proxy-miniature/pwm-motor/src/PwmMotor.cpp]: q: how does getModuleStateAndWaitForRemainingTimeInTimeslice() work? e.g. how many iterations of body() with 1Hz ?

Comment (bjornborg) [inline code/proxy-miniature/pwm-motor/src/PwmMotor.cpp]: It ensure a software realtime runtime. http://opendavinci.cse.chalmers.se/api/classodcore_1_1base_1_1module_1_1AbstractModule.html#a5b63aa67022c4f985cf81cb4f8931a80

File: code/proxy-miniature/pwm-motor/src/PwmMotor.cpp (hunk 8 above)
Comment (bjornborg) [inline code/proxy-miniature/pwm-motor/src/PwmMotor.cpp]: What's the question?

File: code/proxy-miniature/pwm-motor/man/opendlv-proxy-miniature-pwm-motor.1 (hunk 7 above)
Comment (bjornborg) [inline code/proxy-miniature/pwm-motor/man/opendlv-proxy-miniature-pwm-motor.1]: Manual, for execution. This should be filled out more thoroughly when the module is finished.

File: code/proxy-miniature/pwm-motor/src/PwmMotor.cpp (hunk 6 above)
Comment (bjornborg) [inline code/proxy-miniature/pwm-motor/src/PwmMotor.cpp]: Yes, sometimes you want to debug the runtime. This is done through a param.

File: resources/odvd/ODVDMiniature.odvd (hunk 5 above)
Comment (bjornborg) [inline resources/odvd/ODVDMiniature.odvd]: Defining a new container message for internal communication within the conference.

File: usecases/latest/test.pwmmotor/docker-compose.yml (hunk 4 above)
Comment (bjornborg) [inline usecases/latest/test.pwmmotor/docker-compose.yml]: Default is on 1 hertz.

File: code/proxy-miniature/pwm-motor/src/PwmMotor.cpp (hunk 2 above)
Comment (bjornborg) [inline code/proxy-miniature/pwm-motor/src/PwmMotor.cpp]: A way to set the motor on standby, e.g. idle.

File: usecases/latest/test.pwmmotor/configuration (hunk 3 above)
Comment (bjornborg) [inline usecases/latest/test.pwmmotor/configuration]: The configuration may be discarded/ignored. Not being used.

**Comment by bjornborg** [comment]
Ping @Remsworn

File: code/proxy-miniature/pwm-motor/src/PwmMotor.cpp (hunk 8 above)
Comment (Remsworn) [inline code/proxy-miniature/pwm-motor/src/PwmMotor.cpp]: Should m_initialised be set to false in PwmMotor::tearDown() ?

File: code/proxy-miniature/pwm-motor/src/PwmMotor.cpp [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+#include <odvdminiature/GeneratedHeaders_ODVDMiniature.h>
+
+#include "PwmMotor.h"
+
+namespace opendlv {
+namespace proxy {
+namespace miniature {
+
+PwmMotor::PwmMotor(const int &argc, char **argv)
+    : TimeTriggeredConferenceClientModule(argc, argv, "proxy-miniature-pwm-motor")
```
Comment (Remsworn) [inline code/proxy-miniature/pwm-motor/src/PwmMotor.cpp]: vad innebär rad 40-57?

Comment (bjornborg) [inline code/proxy-miniature/pwm-motor/src/PwmMotor.cpp]: General constructs and deconstructors in general object oriented programming.

File: code/proxy-miniature/pwm-motor/src/PwmMotor.cpp (hunk 8 above)
Comment (bjornborg) [inline code/proxy-miniature/pwm-motor/src/PwmMotor.cpp]: The variable is set to false during initiation and becomes true in setup function. This is due to multithread problems if Next obtained gets invoked and functions are dependent on initiation of member variables. By having this flag, you will have a controlled initiation and runtime.

**Pull request merged**